FfiSpike/GDBHelper.cpp

```diff
@@ -16,31 +20,111 @@ std::string get_gdb_message(std::string message)
 	return result.str();
 }
 
+#define RETURN_IF_FALSE(expr) ; if(!(expr)) { return false; }
+
 int gdb_send_message(std::string message, SOCKET socket, long long length)
 {
 	return send_message(get_gdb_message(message), socket, length);
 }
 
-void init(std::string& executable, SOCKET socket)
+bool await_response(std::string message, SOCKET socket, std::string expected_response = kGDBOk)
+{
+	gdb_send_message(message, socket);
+	std::string answer = receive_message_raw(socket);
+	bool matches_expected_response = contains(answer, expected_response) || contains(answer, kGDBOk);
+	return matches_expected_response && !starts_with(answer, kGDBErrorPrefix);
+}
+
+bool init(std::string& executable, SOCKET socket, std::string& application_identifier, std::map<std::string, ApplicationCache>& apps_cache)
 {
-	gdb_send_message("QStartNoAckMode", socket);
+	if (apps_cache.count(application_identifier) && apps_cache[application_identifier].has_initialized_gdb)
+		return true;
+	// Noramlly GDB requires + or - acknowledgments after every message
+	// This however is only valuable if the communication channel is insecure (packets are lost - e.g. UDP)
+	// In our case this would only cause overhead so we disable it using QStartNoAckMode
+	RETURN_IF_FALSE(await_response("QStartNoAckMode", socket, "+"));
 	send_message("+", socket);
-	gdb_send_message("QEnvironmentHexEncoded:", socket);
-	gdb_send_message("QSetDisableASLR:1", socket);
+	// Set QEnvironmentHexEncoded because the arguments we pass around (e.g. path to executable) may contain non-printable characters
+	RETURN_IF_FALSE(await_response("QEnvironmentHexEncoded:", socket));
+	// This actually means QSetDisableASLR:TRUE and actually disables Address space layout randomization on the next A request
+	// We need this because GDB isn't connected to any inferior process (e.g. the application we want to control)
+	RETURN_IF_FALSE(await_response("QSetDisableASLR:1", socket));
 	std::stringstream result;
+	// Initializes the argv in the form of arglen,argnum,arg
+	// In our case:
+	// arglen - twice the size of the original argument because we HEX encode it and every symbol becomes two bytes
+	// argnum - 0 we only have one arg
+	// arg - the actual arg, which is HEX encoded because we set the environment that way
 	result << "A" << executable.size() * 2 << ",0," << to_hex(executable);
-	gdb_send_message(result.str(), socket);
+	RETURN_IF_FALSE(await_response(result.str(), socket));
+	// After all of this is done we can actually call a method with these arguments
+	apps_cache[application_identifier].has_initialized_gdb = true;
```

Comment:
I think if the user just want to start an application the `apps_cache` will not have `apps_cache[application_identifier]` and `apps_cache[application_identifier].has_initialized_gdb` will throw an error. Can you add check for `apps_cache[application_identifier]` if I am right.